### ISSUE TITLE
Unable to provide pre-calculated linkage for the split heatmap.

### ISSUE THREAD

**Issue by liuzj039** [Marsilea-viz/marsilea#16]
I would like to express my sincere gratitude to the developes.
Recently, I have been experimenting with calling the previously calculated linkage on a heatmap after splitting it. For more detailed information, please refer to the following code.
```
ar_test = np.random.random(size=(10, 10))

# The following code represents the internally calculated linkage by marsilea, and it functions effectively
h = ma.Heatmap(ar_test)
h.hsplit(labels=['a', 'b'] * 5)
h.vsplit(labels=['a', 'b'] * 5)
h.add_dendrogram('left', method='ward')
h.render()

# However, when attempting to use the previously calculated linkage, an error occurs
h1 = ma.Heatmap(ar_test)
h1.hsplit(labels=['a', 'b'] * 5)
h1.vsplit(labels=['a', 'b'] * 5)
h1.add_dendrogram('left', linkage=[x.Z for x in h._deform.get_row_dendrogram().dens])
h1.render()
```

I believe that the issue lies in [here ](https://github.com/Marsilea-viz/marsilea/blob/5c2f64d0b921cbe3ca14a2d9f8bb1291c02781c7/marsilea/_deform.py#L176C29-L176C29) that when a list is provided, where each element represents the linkage of a chunk, it triggers this error. It may be possible to resolve this problem by implementing an additional check. However, ensuring that the provided linkage corresponds to the order of the chunks presents another challenge.

Once again, I extend my heartfelt appreciation to the developer. This package has successfully resolved numerous complex issues I encountered.

**Comment by Mr-Milk** [comment]
May I ask what's the scenario for your use case? Because the `Deformation` is not a public API, it shouldn't be used by user in most cases. But I'll look into the issue, thanks for reporting this issue.

**Comment by liuzj039** [comment]
Thank you very much for your prompt response, and I apologize for not providing detailed use case yesterday.

 I have a type of data that is not a symmetric matrix, but there is correlation between the rows and columns.
```
h = ma.Heatmap(ar_test, cmap='Reds')
h.render()
```
![image](https://github.com/Marsilea-viz/marsilea/assets/55271666/56b148bf-e272-408e-81af-477029ada41c)

As a result, when I attempt to cluster it, the outcome becomes as follows due to its lack of symmetry.
```
h = ma.Heatmap(ar_test, cmap='Reds')
h.add_dendrogram('left')
h.add_dendrogram('top')
h.render()
```
![image](https://github.com/Marsilea-viz/marsilea/assets/55271666/443ccc8f-9a9c-4daa-bb17-bf8387da4b25)

One solution is to provide the precomputed linkage when performing clustering without spliting the rows and columns.
```
import scipy
h = ma.Heatmap(ar_test, cmap='Reds')
linkage = scipy.cluster.hierarchy.linkage(ar_test)
h.add_dendrogram('left', linkage=linkage)
h.add_dendrogram('top', linkage=linkage, add_base=False, add_meta=False)
h.render()
```
![image](https://github.com/Marsilea-viz/marsilea/assets/55271666/dc8295f7-494b-40fc-a7ec-0e0f3eef27b6)
However as I described above, this becomes unfeasible when splitting the rows and columns. The solution I am currently using is to read the internal linkage of Marsilea and rearrange the data.
```
def changeDataColOrder(ls_orgOrder, ls_newOrder):
    # orgOrder: [0, 2, 4, 6, 8, 10, 12, 14, 16, 1, 3, 5, 7, 9, 11, 13, 15, 17]
    # newOrder: [[7, 3, 5, 1, 4, 0, 6, 2, 8], [6, 8, 5, 3, 4, 1, 7, 0, 2]]
    _ls = []
    i = 0
    for x in ls_newOrder:
        _ls.append(ls_orgOrder[i: i+len(x)])
        i += len(x)
    ls_orgOrder = _ls

    ls_final = []
    for ls_x, ls_y in zip(ls_orgOrder, ls_newOrder):
        for y in ls_y:
            ls_final.append(ls_x[y])
    return ls_final

h = ma.Heatmap(ar_test, cmap='Reds')
h.hsplit(labels=['a', 'b'] * 9)
h.vsplit(labels=['a', 'b'] * 9)
# h.add_dendrogram('left', method='ward')
# h.add_dendrogram('top', method='ward')
h.add_dendrogram('left', method='ward')

# h._col_den = h._row_den
h.render()
plt.close()
h._deform.set_data_col_reindex(changeDataColOrder(h._deform.data_row_reindex, h._deform.row_reorder_index))
h.render()
```
![image](https://github.com/Marsilea-viz/marsilea/assets/55271666/be33ec7d-3616-4218-b1bb-ad041c02cc1b)
Thanks again for your quick reply.

**Comment by Mr-Milk** [comment]
Your use case makes sense to me, I will definitely look into it. But I'll need to think about what's the best way to design API for this use case and it doesn't seem easy to implement (Which means it won't be fixed soon, sorry!). I will plan it as part of the next minor release. Feel free to comment if you have any ideas in mind.

**Comment by liuzj039** [comment]
Very much looking forward to the updates!

**Comment by liuzj039** [comment]
I am very grateful for your dedication to this issue. I believe providing a dictionary directly is an excellent solution. But I still have a few concerns about the order of the divided matrices. It would be manageable if the list is of the type like ['a', 'a', 'a', 'b', 'b', 'b'], but if it is of the type ['a', 'b', 'a', 'b', 'a', 'b'], there might be potential differences between the matrix order used by the user for calculating the linkage and the Marsilea. Nevertheless, I believe it is necessary to take these risks into account as this is not a very typical use case. To address this issue, it would be ideal if we could extract the internally splited matrices.But it appears that the matrics is trully splited during rendering (I apologize if it is incorrect). I am unsure if implementing the extraction of the splited matrices would be meaningful. Additionally, I consider it a very good strategy to always set add_meta as False when providing linkage.

**Comment by Mr-Milk** [comment]
You are right, everything in Marsilea is lazy until the final rendering. Because we try not to force users to call `.add_dendrogram()` and `.vsplit()` or `.hsplit()` in order, they can call in any order they want. You can even call `.add_dendrogram()` multiple times to add it at different places. Given such consideration, we can only know whether a dendrogram should be split until rendering.

It's possible to have a solution as below, this can ensure the order is correct no matter how meta dendrogram changes. You have `add_meta=True` 😊. But this will need you to render the plot twice.

```python
h = ma.Heatmap(data)
h.vsplit(labels=['a', 'b', 'a', b', 'a', 'b'], order=['a', 'b'])
h.add_dendrogram('top')
h.render()

col_linkage = h.get_col_linkage()
# Something like {'a': Z1, 'b': Z1}
# If no label, it would be index: {0: Z1, 1: Z2}

# And then you can rerender it like
h = ma.Heatmap(data)
h.vsplit(labels=['a', 'b', 'a', b', 'a', 'b'], order=['a', 'b'])
h.hsplit(labels=['a', 'b', 'a', b', 'a', 'b'], order=['a', 'b'])
h.add_dendrogram('top')
h.add_dendrogram('left', linkage=col_linkage)
h.render()
```

Feel free to comment! If this is good then I will try to start the implementation. Meanwhile, it's more than welcome if you want to submit a pull request.

**Comment by Mr-Milk** [comment]
Fixed by #17 #18
